### PR TITLE
feature: add lua_ssl_conf_command directive for setting arbitrary Ope…

### PR DESCRIPTION
**Pull request by catbro666**
…nSSL configuration parameter particularly the TLSv1.3 ciphersuites

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

File: README.markdown [hunk 1]
```diff
@@ -3138,6 +3139,34 @@ See also [lua_ssl_trusted_certificate](#lua_ssl_trusted_certificate).
...
+The directive is supported when using OpenSSL 1.0.2 or higher and nginx 1.19.4 or higher. According to the specify command, higher OpenSSL version may be needed.
+
+Several `lua_ssl_conf_command` directives can be specified on the same level:
+
+```
+ssl_conf_command Options PrioritizeChaCha;
+ssl_conf_command Ciphersuites TLS_CHACHA20_POLY1305_SHA256;
+```
+
+Configuration commands are applied after lua-nginx own configuration for SSL, so they can be used to override anything set by lua-nginx.
```
Comment (spacewander) [inline README.markdown:3162]: lua-nginx => OpenResty

File: README.markdown [hunk 2]
```diff
@@ -3138,6 +3139,34 @@ See also [lua_ssl_trusted_certificate](#lua_ssl_trusted_certificate).
...
+Several `lua_ssl_conf_command` directives can be specified on the same level:
+
+```
+ssl_conf_command Options PrioritizeChaCha;
+ssl_conf_command Ciphersuites TLS_CHACHA20_POLY1305_SHA256;
+```
+
+Configuration commands are applied after lua-nginx own configuration for SSL, so they can be used to override anything set by lua-nginx. 
+
+Note though that configuring OpenSSL directly with `lua_ssl_conf_command` might result in a behaviour nginx does not expect, and should be done with care.
```
Comment (spacewander) [inline README.markdown:3164]: nginx => OpenResty

File: README.markdown [hunk 3]
```diff
@@ -3138,6 +3139,34 @@ See also [lua_ssl_trusted_certificate](#lua_ssl_trusted_certificate).
...
+
+This directive was first introduced in the `v0.10.20` release.
```
Comment (spacewander) [inline README.markdown:3166]: Should be `v0.10.21`?

File: t/129-ssl-socket.t [hunk 4]
```diff
@@ -2561,3 +2561,275 @@ qr/\[error\] .* ngx.socket sslhandshake: expecting 1 ~ 5 arguments \(including t
...
+                ngx.exit(200)
+            }
+        }
+    }
+--- config
+    server_tokens off;
+    lua_ssl_protocols TLSv1.3;
+
+    location /t {
+        content_by_lua '
```
Comment (spacewander) [inline t/129-ssl-socket.t:2587]: Please use the content_by_lua_block style

File: README.markdown (hunk 3 above)
Comment (catbro666) [reply README.markdown]: I‘m not sure about it. I noticed the current version is `v0.10.20rc1`, so I thought the next release version would be `v0.10.20`..

Comment (catbro666) [reply README.markdown]: I‘m not sure about it. I noticed the current version is `v0.10.20rc1`, so I thought the next release version would be `v0.10.20`..

File: README.markdown [hunk 5]
```diff
@@ -3138,6 +3139,34 @@ See also [lua_ssl_trusted_certificate](#lua_ssl_trusted_certificate).
...
+**context:** *http, server, location*
+
+Sets arbitrary OpenSSL configuration [commands](https://www.openssl.org/docs/man1.1.1/man3/SSL_CONF_cmd.html).
+
+The directive is supported when using OpenSSL 1.0.2 or higher and nginx 1.19.4 or higher. According to the specify command, higher OpenSSL version may be needed.
+
+Several `lua_ssl_conf_command` directives can be specified on the same level:
+
+```
+ssl_conf_command Options PrioritizeChaCha;
```
Comment (spacewander) [inline README.markdown:3158]: Should be lua_ssl_conf_command

File: src/ngx_http_lua_module.c [hunk 6]
```diff
@@ -1330,6 +1350,15 @@ ngx_http_lua_merge_loc_conf(ngx_conf_t *cf, void *parent, void *child)
         return NGX_CONF_ERROR;
     }
 
+#if (nginx_version >= 1019004)
+    ngx_conf_merge_ptr_value(conf->ssl_conf_commands, prev->ssl_conf_commands,
+                             NULL);
+    if (ngx_ssl_conf_commands(cf, conf->ssl, conf->ssl_conf_commands)
```
Comment (spacewander) [inline src/ngx_http_lua_module.c:1356]: IMHO, it would be easier to read if we put this into `ngx_http_lua_set_ssl`.

File: README.markdown [hunk 7]
```diff
@@ -1153,6 +1153,7 @@ Directives
 * [lua_ssl_protocols](#lua_ssl_protocols)
 * [lua_ssl_trusted_certificate](#lua_ssl_trusted_certificate)
 * [lua_ssl_verify_depth](#lua_ssl_verify_depth)
+* [lua_ssl_conf_command](#lua_ssl_conf_command)
```
Comment (spacewander) [inline README.markdown:1156]: Need to sync the change to https://github.com/openresty/lua-nginx-module/blob/master/doc/HttpLuaModule.wiki

**Review by spacewander — APPROVED** [review]

**Review by doujiang24 — APPROVED** [review]

**Pull request merged**